### PR TITLE
Do not reset filters when saving, cancelling, refreshing dashboards periodically

### LINKED ISSUE #5332: Don't change parameter values when leaving edit dashboard mode
### Feature requests and proposals

I am not sure if this is a regression or just an oversight (I am on `0.24.2`), but when leaving the dashboard edit mode, the parameters get changed, which is quite annoying, because when you are setting multiple boards/cards you have to fill in all the values all the time.

When you leave the edit mode either with `save` or `cancel`, then filters without default values are removed completely and filters with default value are reset to this value. This is probably due to removing the parameters from the URL, but they are there in the edit mode, they disappear only when leaving it.

With `cancel` this was a little bit more unpredictable, sometimes the value without a default would stay filled in, haven't found the reason behind the difference in the behavior.

This might have been partly intentional, because when saving the dashboard, the parameters could have changed, so removing them might be ensuring, there are not filled in any incompatible values, but if this is the case and it is really needed, then this should not be the case for `cancel` and with `save` I think a change of type of the particular filter should be detected first before removing it.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/5332

Preserves filters in mentioned cases and validates they still exist.

How test test manually:
1. Go to any dashboard that already has the dashboard filter connected to any of the fields
2. Set the filter value if it already wasn't set
3. Note that the url changes and it now includes the filter and its value (correct behavior)
4. Click on the pencil icon to enter Dashboard edit mode
5. Click on "Cancel" button
6. The filter should not be changed

### CODE REVIEW

**Review by alxnddr — COMMENTED** [review]
@ranquild looks good and nicely works when users `cancel` editing a dashboard. However, it still resets the filter values on `save`

File: frontend/src/metabase/dashboard/dashboard.js [hunk 1]
```diff
@@ -663,7 +661,7 @@ function expandInlineCard(card) {
 export const fetchDashboard = createThunkAction(FETCH_DASHBOARD, function(
   dashId,
   queryParams,
-  enableDefaultParameters = true,
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/dashboard.js:666]: Never passed false

File: frontend/src/metabase/dashboard/dashboard.js [hunk 2]
```diff
@@ -712,13 +710,17 @@ export const fetchDashboard = createThunkAction(FETCH_DASHBOARD, function(
...
+    const newParameterValues = {};
+
     if (result.parameters) {
       for (const parameter of result.parameters) {
-        if (queryParams && queryParams[parameter.slug] != null) {
-          parameterValues[parameter.id] = queryParams[parameter.slug];
-        } else if (enableDefaultParameters && parameter.default != null) {
-          parameterValues[parameter.id] = parameter.default;
+        if (preserveParameters) {
+          newParameterValues[parameter.id] = oldParameterValues[parameter.id];
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/dashboard.js:719]: It's important that we only preserve parameters that exist in the response

**Comment by ranquild** [comment]
> @ranquild looks good and nicely works when users `cancel` editing a dashboard. However, it still resets the filter values on `save`

Fixed, please have a look

**Review by alxnddr — APPROVED** [review]
👍

**Comment by ranquild** [comment]
@metabase-bot backport release-x.41.x

**Comment by ranquild** [comment]
@metabase-bot backport release-x.40.x

**Pull request merged**